### PR TITLE
Refactor HPMC walls and port to v3

### LINKED ISSUE #878: Refactor HPMC external fields and port to v3 API
## Description

Port needed HPMC external fields to v3. We should also refactor the internals of how these fields are implemented as the code is very complex and hard to follow.

## Motivation and context

This issue is closely related with #551 which deals specifically with the walls and crosses over to MD as well.

As for the non-walls, we need to consider:
* hpmc.field.callback - What problem does this solve? Would it be better to provide a different interface for custom external potentials? Certainly in v3 it should be possible to leverage local snapshots.
* hpmc.field.external_field_composite - We should remove this and consider alternatives (i.e. HPMCIntegrator could store a list of external fields).
* hpmc.field.frenkel_ladd_energy - We should remove this. It is just a shortcut for lattice_field + remove_drift. Users should configure and manage those separately if they wish to.
* hpmc.field.lattice_field - Needed for frenkel-ladd compuations.

### PR DESCRIPTION
**Pull request by tcmoore3**
## Description

See title.

## Motivation and context

This PR addresses HPMC walls, which is the last item to be addressed in #878, so this PR therefore resolves #878. 

## How has this been tested?

I have run several validation tests on systems of hard octahedra with various walls and combinations of walls:

1. Sphere wall at origin with `inside = True` and separately with `inside = False` 
2. Cylinder wall oriented along z-axis with `inside = True` and separately with `inside = False`
3. Plane wall with `normal = (0, 0, 1)` and separately with `normal = (0, 0, -1)` and separately with `normal = (1, 0, 0)`
4. A combination of a sphere wall (with `inside = True`) and a plane wall with `normal = (0, 0, 1)` to confine particles to a hemisphere

I observe the expected behavior in all cases.

## Change log

```
- Refactor HPMC walls and port to v3 API
```

## Running to-do list

- [ ] Add unit tests

## Checklist:

- [x] I have reviewed the [**Contributor Guidelines**](https://github.com/glotzerlab/hoomd-blue/blob/master/CONTRIBUTING.md).
- [x] I agree with the terms of the [**HOOMD-blue Contributor Agreement**](https://github.com/glotzerlab/hoomd-blue/blob/master/ContributorAgreement.md).
- [x] My name is on the list of contributors (`sphinx-doc/credits.rst`) in the pull request source branch.

### CODE REVIEW

**Review by tommy-waltmann — CHANGES_REQUESTED** [review]
Good work so far, this seems to be getting pretty close

File: hoomd/hpmc/external/wall.py [hunk 1]
```diff
@@ -1,525 +1,86 @@
...
+            hoomd.wall.Sphere: self._cpp_obj.SphereWalls,
+            hoomd.wall.Cylinder: self._cpp_obj.CylinderWalls,
+            hoomd.wall.Plane: self._cpp_obj.PlaneWalls,
+        })
+        super()._attach()
+
+    @property
+    def walls(self):
+        """`list` [`hoomd.wall.WallGeometry`]: \
+            The walls associated with this wall potential."""
```
Comment (joaander) [inline hoomd/hpmc/external/wall.py:79]:
```suggestion
            The wall geometries associated with this potential."""
```

**Review by joaander — CHANGES_REQUESTED** [review]
Nice implementation. @b-butler's work on the meta list paid off.

I have some comments for improvement, especially in the documentation.

File: hoomd/hpmc/ExternalFieldHarmonic.h [hunk 2]
```diff
@@ -463,9 +463,6 @@ template<class Shape> void export_HarmonicField(pybind11::module& m, std::string
                       &ExternalFieldHarmonic<Shape>::setSymmetricallyEquivalentOrientations)
         .def("getEnergies", &ExternalFieldHarmonic<Shape>::getEnergies);
     }
-
-void export_HarmonicFields(pybind11::module& m);
-
```
[the remark this answers is not in the document]
Comment (tcmoore3) [reply hoomd/hpmc/ExternalFieldHarmonic.h:468]: Just general cleanup. It should have been in #1177 but I just missed it there.

File: hoomd/hpmc/module_convex_polygon.cc [hunk 3]
```diff
@@ -45,8 +44,6 @@ void export_convex_polygon(pybind11::module& m)
     export_HarmonicField<ShapeConvexPolygon>(m, "ExternalFieldHarmonicConvexPolygon");
     export_ExternalFieldComposite<ShapeConvexPolygon>(m, "ExternalFieldCompositeConvexPolygon");
     // export_ExternalFieldWall<ShapeConvexPolygon>(m, "WallConvexPolygon");
```
[the remark this answers is not in the document]
Comment (tcmoore3) [reply hoomd/hpmc/module_convex_polygon.cc]: The original developers did not implement overlap checks for anisotropic 2D shapes. I don't know why this was commented out instead of raising an error if for unsupported shape–wall combinations. I have fixed that now.

File: hoomd/hpmc/external/wall.py [hunk 4]
```diff
@@ -61,6 +61,26 @@ def _attach(self):
         if not isinstance(integrator, hoomd.hpmc.integrate.HPMCIntegrator):
             raise RuntimeError('Walls require a valid HPMC integrator.')
 
+        # check that shape-wall overlap checks are implemented, error if not
+        supported_shape_wall_overlap_checks = {
```
Comment (joaander) [inline hoomd/hpmc/external/wall.py:65]: This will check the that the wall/shape combo is supported at initialization time. If possible, also check on setting new walls to error in cases where the user starts with a supported combo and then assigns an unsupported one. This could occur for example when starting with an empty wall list and a `ConvexSpheropolyhedron` integrator, then adding a cylinder wall.

File: hoomd/hpmc/ExternalFieldWall.cc [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+namespace hoomd
+    {
+namespace hpmc
+    {
+namespace detail
+    {
+void export_wall_classes(pybind11::module& m)
+    {
+    pybind11::class_<hoomd::hpmc::SphereWall>(m, "SphereWall")
+        .def(pybind11::init(
```
[the remark this answers is not in the document]
Comment (tcmoore3) [reply hoomd/hpmc/ExternalFieldWall.cc]: I (along with Tommy) made this change in [this commit](https://github.com/glotzerlab/hoomd-blue/pull/1216/commits/92eca28f00211fe749f67bcb887c3d60f299a330).

File: hoomd/hpmc/external/wall.py [hunk 6]
```diff
@@ -1,525 +1,111 @@
...
+            ],
+            hoomd.hpmc.integrate.ConvexPolyhedron: [
+                hoomd.wall.Sphere, hoomd.wall.Cylinder, hoomd.wall.Plane
+            ],
+            hoomd.hpmc.integrate.ConvexSpheropolyhedron: [
+                hoomd.wall.Sphere, hoomd.wall.Plane
+            ]
+        }
+        integrator_type = type(integrator)
+        for wt in [type(w) for w in walls]:
```
Comment (b-butler) [inline hoomd/hpmc/external/wall.py:70]:
```suggestion
        for wt in (type(w) for w in walls):
```
Use generator for for-loops.

File: hoomd/hpmc/external/wall.py (hunk 4 above)
Comment (b-butler) [reply hoomd/hpmc/external/wall.py:58]: Given the nature of `_WallsMetaList` we need to change the available types when attaching and remove unavailable types at attaching. Currently this is not a feature from the class so it would need to be monkey patching in a sense here which would involve removing and modifying some attributes of the list.

**Review by b-butler — CHANGES_REQUESTED** [review]
Remove `inside` from `PlaneWall` in C++.

I may have found the segfault error, but I am not sure. I think it may have to due with the way you created the constructors for the wall classes in C++.

File: hoomd/hpmc/external/wall.py (hunk 4 above)
Comment (tcmoore3) [reply hoomd/hpmc/external/wall.py:58]: Good catch. I refactored this check into a separate class method that is used here and in the walls setter.

File: hoomd/hpmc/integrate.py [hunk 7]
```diff
@@ -456,6 +456,11 @@ class Sphere(HPMCIntegrator):
               `True` to ignore tracked statistics.
             * ``orientable`` (`bool`, **default:** `False`) - set to `True` to
               allow rotation moves on this particle type.
+
+    .. rubric:: Wall support.
+
+    `integrate.Sphere` contains support for all `hoomd.wall` geometries.
```
Comment (joaander) [inline hoomd/hpmc/integrate.py:462]:
```suggestion
    `integrate.Sphere` supports all `hoomd.wall` geometries.
```
Use active voice.

File: hoomd/hpmc/integrate.py [hunk 8]
```diff
@@ -551,6 +556,10 @@ class ConvexPolygon(HPMCIntegrator):
               HPMC does not check that all vertex requirements are met.
               Undefined behavior **will result** when they are violated.
 
+    .. rubric:: Wall support.
+
+    `integrate.ConvexPolygon` contains support for no `hoomd.wall` geometries.
```
Comment (joaander) [inline hoomd/hpmc/integrate.py:561]:
```suggestion
    `integrate.ConvexPolygon` supports no `hoomd.wall` geometries.
```

File: hoomd/hpmc/integrate.py [hunk 9]
```diff
@@ -653,6 +662,12 @@ class ConvexSpheropolygon(HPMCIntegrator):
             Warning:
                 HPMC does not check that all vertex requirements are met.
                 Undefined behavior will result when they are violated.
+
+    .. rubric:: Wall support.
+
+    `integrate.ConvexSpheropolygon` contains support for no `hoomd.wall`
```
Comment (joaander) [inline hoomd/hpmc/integrate.py:668]:
```suggestion
    `integrate.ConvexSpheropolygon` supports no `hoomd.wall`
```

File: hoomd/hpmc/integrate.py [hunk 10]
```diff
@@ -749,6 +764,11 @@ class SimplePolygon(HPMCIntegrator):
                 HPMC does not check that all vertex requirements are met.
                 Undefined behavior will result when they are violated.
 
+
+    .. rubric:: Wall support.
+
+    `integrate.SimplePolygon` contains support for no `hoomd.wall` geometries.
```
Comment (joaander) [inline hoomd/hpmc/integrate.py:770]:
```suggestion
    `integrate.SimplePolygon` supports no `hoomd.wall` geometries.
```

File: hoomd/hpmc/integrate.py [hunk 11]
```diff
@@ -918,6 +938,11 @@ class Polyhedron(HPMCIntegrator):
             Warning:
                 HPMC does not check that all vertex requirements are met.
                 Undefined behavior will result when they are violated.
+
+    .. rubric:: Wall support.
+
+    `integrate.Polyhedron` contains support for no `hoomd.wall` geometries.
```
Comment (joaander) [inline hoomd/hpmc/integrate.py:944]:
```suggestion
    `integrate.Polyhedron` supports no `hoomd.wall` geometries.
```

File: hoomd/hpmc/integrate.py [hunk 12]
```diff
@@ -1032,6 +1057,12 @@ class ConvexPolyhedron(HPMCIntegrator):
             Warning:
                 HPMC does not check that all vertex requirements are met.
                 Undefined behavior will result when they are violated.
+
+    .. rubric:: Wall support.
+
+    `integrate.ConvexPolyhedron` contains support for all `hoomd.wall`
```
Comment (joaander) [inline hoomd/hpmc/integrate.py:1063]:
```suggestion
    `integrate.ConvexPolyhedron` supports all `hoomd.wall`
```

File: hoomd/hpmc/integrate.py [hunk 13]
```diff
@@ -1174,6 +1205,12 @@ class FacetedEllipsoid(HPMCIntegrator):
                 specified, the half-space intersection of the normals must match
                 the convex polyhedron defined by the vertices (if non-empty),
                 the half-space intersection is **not** calculated automatically.
+
+    .. rubric:: Wall support.
+
+    `integrate.FacetedEllipsoid` contains support for no `hoomd.wall`
```
Comment (joaander) [inline hoomd/hpmc/integrate.py:1211]:
```suggestion
    `integrate.FacetedEllipsoid` supports no `hoomd.wall`
```

File: hoomd/hpmc/integrate.py [hunk 14]
```diff
@@ -1253,6 +1290,11 @@ class Sphinx(HPMCIntegrator):
               :math:`[\\mathrm{length}]`.
             * ``ignore_statistics`` (`bool`, **default:** `False`) - set to
               `True` to ignore tracked statistics.
+
+    .. rubric:: Wall support.
+
+    `integrate.Sphinx` contains support for no `hoomd.wall` geometries.
```
Comment (joaander) [inline hoomd/hpmc/integrate.py:1296]:
```suggestion
    `integrate.Sphinx` supports no `hoomd.wall` geometries.
```

File: hoomd/hpmc/integrate.py [hunk 15]
```diff
@@ -1347,6 +1389,12 @@ class ConvexSpheropolyhedron(HPMCIntegrator):
             Warning:
                 HPMC does not check that all vertex requirements are met.
                 Undefined behavior will result when they are violated.
+
+    .. rubric:: Wall support.
+
+    `integrate.ConvexSpheropolyhedron` contains support for the
```
Comment (joaander) [inline hoomd/hpmc/integrate.py:1395]:
```suggestion
    `integrate.ConvexSpheropolyhedron` supports the
```

File: hoomd/hpmc/integrate.py [hunk 16]
```diff
@@ -1432,6 +1480,11 @@ class Ellipsoid(HPMCIntegrator):
               direction :math:`[\\mathrm{length}]`
             * ``ignore_statistics`` (`bool`, **default:** `False`) - set to
               `True` to ignore tracked statistics.
+
+    .. rubric:: Wall support.
+
+    `integrate.Ellipsoid` contains support for no `hoomd.wall`  geometries.
```
Comment (joaander) [inline hoomd/hpmc/integrate.py:1486]:
```suggestion
    `integrate.Ellipsoid` supports no `hoomd.wall`  geometries.
```

File: hoomd/hpmc/integrate.py [hunk 17]
```diff
@@ -1542,6 +1595,11 @@ class SphereUnion(HPMCIntegrator):
               particles per leaf node to adjust performance.
             * ``ignore_statistics`` (`bool`, **default:** `False`) - set to
               `True` to ignore tracked statistics.
+
+    .. rubric:: Wall support.
+
+    `integrate.SphereUnion` contains support for no `hoomd.wall`  geometries.
```
Comment (joaander) [inline hoomd/hpmc/integrate.py:1601]:
```suggestion
    `integrate.SphereUnion` supports no `hoomd.wall`  geometries.
```

File: hoomd/hpmc/integrate.py [hunk 18]
```diff
@@ -1672,6 +1730,12 @@ class ConvexSpheropolyhedronUnion(HPMCIntegrator):
               particles per leaf node to adjust performance.
             * ``ignore_statistics`` (`bool`, **default:** `False`) - set to
               `True` to ignore tracked statistics.
+
+    .. rubric:: Wall support.
+
+    `integrate.ConvexSpheropolyhedronUnion` contains support for no `hoomd.wall`
```
Comment (joaander) [inline hoomd/hpmc/integrate.py:1736]:
```suggestion
    `integrate.ConvexSpheropolyhedronUnion` supports no `hoomd.wall`
```

File: hoomd/hpmc/integrate.py [hunk 19]
```diff
@@ -1814,6 +1878,12 @@ class FacetedEllipsoidUnion(HPMCIntegrator):
               particles per leaf node to adjust performance.
             * ``ignore_statistics`` (`bool`, **default:** `False`) - set to
               `True` to ignore tracked statistics.
+
+    .. rubric:: Wall support.
+
+    `integrate.FacetedEllipsoidUnion` contains support for no `hoomd.wall`
```
Comment (joaander) [inline hoomd/hpmc/integrate.py:1884]:
```suggestion
    `integrate.FacetedEllipsoidUnion` supports no `hoomd.wall`
```

File: hoomd/hpmc/pytest/test_external_wall.py [hunk 20]
```diff
@@ -0,0 +1,227 @@
+# Copyright (c) 2009-2022 The Regents of the University of Michigan.
+# Part of HOOMD-blue, released under the BSD 3-Clause License.
+
+"""Test hoomd.hpmc.external.wall."""
```
Comment (joaander) [inline hoomd/hpmc/pytest/test_external_wall.py:4]: Add tests that place particles and check that they overlap (or not overlap) with walls.

**Review by joaander — CHANGES_REQUESTED** [review]
Thanks, this is almost done.

I made some suggestions to use active voice in the documentation.

Add some unit tests that check whether particles overlap with walls correctly.

File: hoomd/hpmc/pytest/test_external_wall.py [hunk 21]
```diff
@@ -0,0 +1,227 @@
...
+@pytest.mark.cpu
+def test_replace_with_valid(simulation_factory, two_particle_snapshot_factory,
+                            add_default_integrator):
+    sim = simulation_factory(two_particle_snapshot_factory())
+    integrator_class = hoomd.hpmc.integrate.ConvexSpheropolyhedron
+    walls = [hoomd.wall.Plane]
+    mc, walls = add_default_integrator(sim, integrator_class, walls)
+    sim.run(0)
+    mc.external_potential.walls = [hoomd.wall.Sphere(1.0)]
+    sim.run(0)
```
Comment (b-butler) [inline hoomd/hpmc/pytest/test_external_wall.py:227]: No need for this line.
```suggestion
```

**Review by b-butler — CHANGES_REQUESTED** [review]
We need to add a  test that checks direct modification of the `_MetaWallList` it will fail currently though.

File: hoomd/hpmc/pytest/test_external_wall.py (hunk 20 above)
Comment (tcmoore3) [reply hoomd/hpmc/pytest/test_external_wall.py:4]: Done

File: hoomd/hpmc/external/wall.py [hunk 22]
```diff
@@ -1,525 +1,152 @@
...
-        if xz is None:
-            xz = self.cpp_compute.GetCurrBoxTiltFactorXZ()
-        if yz is None:
-            yz = self.cpp_compute.GetCurrBoxTiltFactorYZ()
-
-        self.cpp_compute.SetCurrBox(Lx, Ly, Lz, xy, xz, yz)
+        if self._attached:
+            integrator = self._simulation.operations.integrator
+            self._check_valid_shape_wall_pairs(wall_list, integrator)
+        self._walls = hoomd.wall._WallsMetaList(wall_list, _to_hpmc_cpp_wall)
```
[the remark this answers is not in the document]
Comment (tcmoore3) [reply hoomd/hpmc/external/wall.py:146]: Good catch

File: hoomd/hpmc/external/wall.py [hunk 23]
```diff
@@ -1,525 +1,152 @@
...
+        }
+        integrator_type = type(integrator)
+        for wt in (type(w) for w in walls):
+            if wt not in supported_shape_wall_overlap_checks.get(
+                    integrator_type, []):
+                msg = f'Overlap checks between {wt} and {integrator_type} are '
+                msg += 'not supported.'
+                raise NotImplementedError(msg)
+
+    def _attach(self):
```
[the remark this answers is not in the document]
Comment (tcmoore3) [reply hoomd/hpmc/external/wall.py:111]: Do you mean something different than what I implemented with `_check_valid_shape_wall_pairs` here?

[the remark this answers is not in the document]
Comment (joaander) [reply hoomd/hpmc/external/wall.py:111]: If it is too much trouble, we don't need to check for this error. We do document what is supported and what is not.

File: hoomd/hpmc/external/wall.py (hunk 22 above)
[the remark this answers is not in the document]
Comment (tcmoore3) [reply hoomd/hpmc/external/wall.py:146]: @b-butler Making this change in the MD code causes tests to fail. I don't know enough about how the metalist is supposed to work to debug on my own. I'll defer to you, let me know if there's anything you need me to do on that.

File: hoomd/hpmc/external/wall.py (hunk 23 above)
[the remark this answers is not in the document]
Comment (tcmoore3) [reply hoomd/hpmc/external/wall.py:111]: I'm a little confused about the original comment. Is `_check_valid_shape_wall_pairs` defined [here](https://github.com/glotzerlab/hoomd-blue/blob/feature/hpmc-walls/hoomd/hpmc/external/wall.py#L91) and called [a few lines below this comment](https://github.com/glotzerlab/hoomd-blue/blob/feature/hpmc-walls/hoomd/hpmc/external/wall.py#L118) and [in the setter](https://github.com/glotzerlab/hoomd-blue/blob/feature/hpmc-walls/hoomd/hpmc/external/wall.py#L145) not dealing with the case where a user adds a new wall that is incompatible?

While we do document what is supported and what is not, I think we should raise this error since this seems like it could be a fairly easy gotcha since 1) wall support is not all-or-nothing for all shapes (e.g., `ConvexSpheropolyhedra` support spherical and planar but not cylindrical walls) and 2) the visibility in the docs is pretty low, just because of how much information is (necessarily) on that page.

**Review by tommy-waltmann — APPROVED** [review]
If you address the comments that Brandon and Josh have left and this should be good.

File: hoomd/hpmc/external/wall.py (hunk 23 above)
[the remark this answers is not in the document]
Comment (tcmoore3) [reply hoomd/hpmc/external/wall.py:111]: Ah, I see what you mean. Yes, I will need help with that.

File: hoomd/md/external/wall.py [hunk 24]
```diff
@@ -212,6 +212,20 @@ def walls(self, wall_list):
         if self._walls is wall_list:
             return
         self._walls = hoomd.wall._WallsMetaList(wall_list, _to_md_cpp_wall)
+        if self._attached:
```
Comment (joaander) [inline hoomd/md/external/wall.py:215]: Does this fix a bug with MD walls?
If yes, we need to add add a draft change log entry on that fix to the PR description.

File: hoomd/wall.py [hunk 25]
```diff
@@ -229,13 +229,15 @@ class Plane(WallGeometry):
...
+            The normal vector to the plane. The vector will be converted to a
             unit vector.
         open (`bool`, optional):
             Whether to include the surface of the plane in the space. ``True``
             means do not include the surface, defaults to ``True``.
 
-    The normal points away from the active half side.
+    The normal points away from the active half side; in HPMC, this means that
+    the normal points to the side of the plane where particles are allowed to
+    be.
```
Comment (joaander) [inline hoomd/wall.py:240]:
```suggestion
    The normal points toward the points with a positive signed distance to the plane.
```
This was a leftover from the v2 documentation. The v3 documentation doesn't use the term "active" anywhere else - I defined everything in terms of the signed distance d. The HPMC documentation also references this.

**Review by joaander — CHANGES_REQUESTED** [review]
Thanks! I have one documentation improvement to make and one question to answer before merging.

File: hoomd/md/external/wall.py (hunk 24 above)
Comment (tcmoore3) [reply hoomd/md/external/wall.py:215]: Yes. It would have led to inconsistent wall lists between the python and c++ code if the walls setter was used to modify the walls in the simulation.

**Review by joaander — APPROVED** [review]
Thanks!

@b-butler also needs to approve.

File: hoomd/hpmc/pytest/test_external_wall.py [hunk 26]
```diff
@@ -222,6 +222,34 @@ def test_replace_with_invalid(simulation_factory, two_particle_snapshot_factory,
...
+                                        two_particle_snapshot_factory,
+                                        add_default_integrator):
+    sim = simulation_factory(two_particle_snapshot_factory())
+    integrator_class = hoomd.hpmc.integrate.ConvexSpheropolyhedron
+    walls = [hoomd.wall.Sphere, hoomd.wall.Plane]
+    mc, walls = add_default_integrator(sim, integrator_class, walls)
+    sim.run(0)
+    with pytest.raises(NotImplementedError):
+        new_walls = [hoomd.wall.Cylinder(1.2345, (0, 0, 0))]
+        mc.external_potential.walls.append(new_walls)
```
Comment (b-butler) [inline hoomd/hpmc/pytest/test_external_wall.py:250]:
```suggestion
        mc.external_potential.walls.extend(new_walls)
```

File: hoomd/hpmc/external/wall.py [hunk 27]
```diff
@@ -21,6 +22,59 @@ def _to_hpmc_cpp_wall(wall):
...
+    def insert(self, index, wall):
+        self._check_wall_compatibility(wall)
+        super().insert(index, wall)
+
+    def append(self, wall):
+        self._check_wall_compatibility(wall)
+        super().append(wall)
+
+
 class WallPotential(ExternalField):
```
[the remark this answers is not in the document]
Comment (tcmoore3) [reply hoomd/hpmc/external/wall.py:78]: I will let you document that in the location and with the language you think is most appropriate.

**Review by b-butler — APPROVED** [review]
